File: validation/check_esp.py
#!/usr/bin/env python3
"""Compare our electrostatic potential integrals against PySCF's.

    ./build/check_esp && python3 validation/check_esp.py

Elementwise over every AO pair at every point, not on a contracted potential.
libcint's grid integral nests the grid index *fastest*, inside the shell pair --
the opposite of its multipole blocks, where the component is slowest -- so a
transposed unpack produces a matrix that is right at one point and wrong at the
rest. Contracting first would average that into a single number that could easily
look plausible.
"""

import pathlib
import sys

import numpy as np

REPO = pathlib.Path(__file__).resolve().parent.parent
sys.path.insert(0, str(REPO / "tools" / "cpu_validation"))
from gen_cpu_validation import bse_to_pyscf, molecule_form, CARTESIAN  # noqa: E402

ATOMS = [("O", (0.0, 0.0, 0.0)),
         ("H", (0.0, 0.0, 0.9584)),
         ("H", (0.9268, 0.0, -0.2400))]

#: **Do not add a Pople basis with sp shells to this comparison without permuting.**
#: PySCF orders a Cartesian 6-31G* oxygen as `1s 2s 3s 2px..2pz 3px..3pz d`, grouping
#: all s functions then all p. Our reader hands libcint the s and p of each shared-
#: exponent group adjacently, giving `1s 2s 2px..2pz 3s 3px..3pz d`, which is also
#: GAMESS's order. So an elementwise comparison of AO-indexed matrices against PySCF
#: is only valid for a basis without sp shells -- the cases here are all like that,
#: which is why they pass.
#:
#: This was found the hard way: using PySCF's overlap with our own localized orbitals
#: reported them as non-orthonormal (`c^T S c` = 0.97, 0.75, 0.74, 0.43) when
#: `check_localize` verifies orthonormality to 2e-15 against our own overlap. The
#: metric was in the wrong order, not the orbitals.

#: Both codes call the same libcint routine on the same basis, so this is the
#: floor set by our BSE table against PySCF's own, not a physics tolerance.
TOL = 1e-8


def main():
    from pyscf import gto

    path = pathlib.Path("/tmp/mqc_esp.txt")
    if not path.exists():
        print(f"  MISSING {path} -- run ./build/check_esp first")
        return 1
    values = [float(x) for x in path.read_text().split()]
    nao, n_grid = int(values[0]), int(values[1])
    at = 2
    grids = np.array(values[at:at + 3*n_grid]).reshape((3, n_grid), order="F")
    at += 3*n_grid
    ours = np.array(values[at:at + nao*nao*n_grid]).reshape((nao, nao, n_grid),
                                                            order="F")

    symbols = {a[0] for a in ATOMS}
    mol = gto.Mole()
    mol.atom = ATOMS
    mol.unit = "Angstrom"
    mol.basis = {s: bse_to_pyscf("cc-pvdz", s) for s in symbols}
    mol.cart = molecule_form("cc-pvdz", symbols) == CARTESIAN
    mol.verbose = 0
    mol.build()
    theirs = mol.intor("int1e_grids", grids=grids.T)      # (n_grid, nao, nao)

    if theirs.shape != (n_grid, nao, nao):
        print(f"  FAIL PySCF returned {theirs.shape}, expected "
              f"{(n_grid, nao, nao)}")
        return 1

    worst = max(np.abs(ours[:, :, g] - theirs[g]).max() for g in range(n_grid))
    print(f"  cc-pvdz  nao {nao}  points {n_grid}")
    print(f"        worst element over {n_grid*nao*nao} integrals: {worst:.2e}")
    for g in range(n_grid):
        gap = np.abs(ours[:, :, g] - theirs[g]).max()
        print(f"        ({grids[0, g]:6.2f},{grids[1, g]:6.2f},{grids[2, g]:6.2f})"
              f"   {gap:.2e}")

    print()
    if worst >= TOL:
        print("[esp] the ESP integrals do not match PySCF's")
        return 1
    print("[esp] our electrostatic potential integrals match PySCF's")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: validation/check_esp.f90
!! Manual check of the electrostatic potential integrals
!!
!!     cmake -B build -DMQC_ENABLE_LIBCINT=ON
!!     ./build/check_esp
!!     python3 validation/check_esp.py
!!
!! `<chi_u| 1/|r - R_g| |chi_v>` at a list of points. Contracted with a density this
!! is the electronic potential the molecule makes, and it is what a charge-
!! penetration screening fit is fitted to: GAMESS optimizes its damping exponents by
!! matching a damped classical multipole potential to the quantum one on a grid, so
!! this integral is the target. The CPU path had no way to compute it -- only the
!! cuEST backend did, through its PCM plan -- so this is the enabling piece for that
!! milestone, and for a CPU polarizable-continuum model later.
!!
!! **Checked elementwise, not on a contracted potential.** libcint runs the grid
!! index fastest and the shell pair outside it, the opposite nesting from the
!! multipole blocks, and a transposed unpack gives a matrix that is plausible at one
!! point and wrong at the others. A single contracted number could hide that; every
!! element at every point cannot.
program check_esp
   use pic_types, only: dp
   use mqc_libcint_integrals, only: libcint_molecule_t, build_libcint_molecule
   use mqc_libcint_rhf, only: rhf_result_t, run_libcint_rhf
   use mqc_libcint_esp, only: esp_matrices
   use mqc_error, only: error_t
   implicit none
   real(dp), parameter :: ANG = 1.8897261254578281_dp

   !> Probe points. Deliberately off-centre, asymmetric, and none on an atom or a
   !> symmetry plane -- a point on either would make several integrals vanish and
   !> hide a transposed unpack.
   integer, parameter :: N_PROBE = 5
   type(libcint_molecule_t) :: mol
   type(rhf_result_t) :: scf
   type(error_t) :: err
   real(dp), allocatable :: v(:, :, :)
   real(dp) :: c(3, 3), g(3, N_PROBE)
   integer :: unit, k
   c = reshape([0.0_dp, 0.0_dp, 0.0_dp, 0.0_dp, 0.0_dp, 0.9584_dp*ANG, &
                0.9268_dp*ANG, 0.0_dp, -0.2400_dp*ANG], [3, 3])
   ! Points off-centre and asymmetric, none on an atom or a symmetry plane.
   g = reshape([1.3_dp, -0.7_dp, 2.1_dp, -2.0_dp, 1.1_dp, 0.4_dp, &
                0.5_dp, 3.0_dp, -1.2_dp, 4.0_dp, 0.2_dp, 0.9_dp, &
                -1.1_dp, -2.2_dp, -3.3_dp], [3, N_PROBE])
   call build_libcint_molecule([8, 1, 1], ["O ", "H ", "H "], c, "cc-pvdz", mol, err)
   call run_libcint_rhf(mol, 10, 200, 1.0e-12_dp, 1.0e-10_dp, .false., scf, err)
   call esp_matrices(mol, g, v, err)
   if (err%has_error()) then
      write (*, *) "FAILED: ", err%get_message()
      stop 1
   end if
   write (*, "(a,3(i0,1x))") " shape: ", size(v, 1), size(v, 2), size(v, 3)
   write (*, "(a)") " electronic potential at each point (a.u.):"
   do k = 1, size(g, 2)
      write (*, "(a,3f8.3,a,f16.10)") "   (", g(:, k), " )  ", -sum(scf%density*v(:, :, k))
   end do
   open (newunit=unit, file="/tmp/mqc_esp.txt", status="replace", action="write")
   write (unit, "(2(i0,1x))") mol%nao, size(g, 2)
   write (unit, "(3es25.16e3)") g
   write (unit, "(es25.16e3)") v
   close (unit)
   call mol%destroy()
end program check_esp
